Add support for notifying the delegate when device configuration changes

nivi-apple opened the pull request:
- Device configuration changes include updates in attributes parts list, server list, device type list, cluster revision or feature map in the descriptor cluster

- Add a test to verify the delegate is notified when device configuration changes

Fixes: #32353

nivi-apple committed:
Add support for notifying the delegate when device configuration changes

- Device configuration changes include updates in attributes parts list, server list,
  device type list, cluster revision or feature map in the descriptor cluster

- Add a test to verify the delegate is notified when device configuration changes

restyled-commits committed:
Restyled by clang-format

nivi-apple committed:
Change device list to device type list

jtung-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1063,6 +1068,18 @@ - (void)_handleReportEnd
 
+    // After the handling of the report, if we detected a device configuration change, notify the delegate
+    // of the same.
+    if (_deviceConfigurationChanged) {
```
You'll want to reset `_deviceConfigurationChanged` to `NO` after this.

nivi-apple replied:
done.

nivi-apple committed:
Add accepted commands list and attribute list to the list of attributes that qualify as device configuration changes

carol-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1107,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
+            // If changes are detected, note that the device configuration has changed.
+            NSDictionary * cachedAttributeDataValue = [self _cachedAttributeValueForPath:attributePath];
+            if (cachedAttributeDataValue != nil && ![self _attributeDataValue:attribute[MTRDataKey] isEqualToDataValue:cachedAttributeDataValue]) {
+                _deviceConfigurationChanged = YES;
```
Jeff left a comment here: https://github.com/project-chip/connectedhomeip/pull/33146/files#r1578444794

jaredgrubb replied:
Ah yep, I just missed it! :heart:

nivi-apple replied:
i just resolved it.

nivi-apple replied:
once we handle report end, i reset it to NO. Jeff had a comment and I fixed it.

nivi-apple committed:
Set _deviceConfigurationChanged to NO in _handleReportEnd

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.h:
```diff
@@ -423,6 +423,16 @@ MTR_EXTERN NSString * const MTRDataVersionKey MTR_NEWLY_AVAILABLE;
 
+/**
+ * Notifies delegate when the device configuration changes. Device configuration changes include updates in parts list, device type list,
+ * server list, accepted commands list, attribute list, feature map or cluster revision attributes in the descriptor cluster.
```
This description is pretty confusing.  I am sure that we are watching for changes to accepted commands list in non-descriptor clusters, yes?

More to the point: should we actually over-specify in the API what "device configuration" means?

nivi-apple replied:
fixed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.h:
```diff
@@ -423,6 +423,16 @@ MTR_EXTERN NSString * const MTRDataVersionKey MTR_NEWLY_AVAILABLE;
+ * Notifies delegate when the device configuration changes. Device configuration changes include updates in parts list, device type list,
+ * server list, accepted commands list, attribute list, feature map or cluster revision attributes in the descriptor cluster.
+ *
+ * This is called when the MTRDevice object detects a change in the device configuration and reports that to the delegate.
```
```suggestion
 * This is called when the MTRDevice object detects a change in the device configuration.
```
This function is obviously "reporting to the delegate".

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.h:
```diff
@@ -423,6 +423,16 @@ MTR_EXTERN NSString * const MTRDataVersionKey MTR_NEWLY_AVAILABLE;
+ *
+ * This is called when the MTRDevice object detects a change in the device configuration and reports that to the delegate.
+ *
+ * The intention is that after this is called, the client should re-enumerate the device topology.
```
```suggestion
 * The intention is that this is called whenever any of the things that usually don't change about a device might have changed.
```

The "device topology" bit is Home terminology that is not related to anything in Matter.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -371,6 +371,11 @@ @implementation MTRDevice {
     NSDate * _Nullable _lastSubscriptionFailureTime;
     MTRDeviceConnectivityMonitor * _connectivityMonitor;
+
+    // This boolean keeps track of any device configuration changes received in an attribute report
+    // and when the report ends, we notify the delegate. Device configuration changes include parts list,
+    // server list, device type list, accepted commands list, attribute list, cluster revision and feature map updates.
```
```suggestion
    // This boolean keeps track of any device configuration changes received in an attribute report.
    // If this is true when the report ends, we notify the delegate.
```

We don't want to have lots of comments that can stop matching the code... ;)

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
 }
 
+// When we receive an attribute report, check if there are any changes in parts list, server list, device type list, accepted commands list,
+//  attribute list, cluster revision or feature map attributes of the descriptor cluster. If yes, make a note that the device configuration changed.
```
Again, the "of the descriptor cluster" does not sound right here.

nivi-apple replied:
fixed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
+    for (NSDictionary<NSString *, id> * attribute in attributeReport) {
+        MTRAttributePath * attributePath = attribute[MTRAttributePathKey];
+
+        if (attributePath.cluster.unsignedLongValue != MTRClusterDescriptorID) {
```
Yeah, so: this check is wrong.

nivi-apple replied:
My bad.this is applicable only for parts list, device type list and server list.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
+        case MTRClusterDescriptorAttributeClusterRevisionID:
+        case MTRClusterDescriptorAttributeFeatureMapID: {
+            // If changes are detected, note that the device configuration has changed.
+            NSDictionary * cachedAttributeDataValue = [self _cachedAttributeValueForPath:attributePath];
```
```suggestion
            MTRDeviceDataValueDictionary cachedAttributeDataValue = [self _cachedAttributeValueForPath:attributePath];
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
+//  attribute list, cluster revision or feature map attributes of the descriptor cluster. If yes, make a note that the device configuration changed.
+- (void)_noteDeviceConfigurationChanged:(NSArray<NSDictionary<NSString *, id> *> *)attributeReport
+{
+    for (NSDictionary<NSString *, id> * attribute in attributeReport) {
```
Should be asserting that lock is held?

nivi-apple replied:
N/A anymore. moved the check to _getAttributesToReportWithReportedValues as suggested as it make sense.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -2974,6 +2974,7 @@ - (void)test031_MTRDeviceAttributeCacheLocalTestStorage
     }
     NSUInteger storedAttributeCountDifferenceFromMTRDeviceReport = dataStoreAttributeCountAfterSecondSubscription - attributesReportedWithSecondSubscription;
     XCTAssertTrue(storedAttributeCountDifferenceFromMTRDeviceReport > 300);
+    [sController removeDevice:device];
```
Please document why this is being done.

nivi-apple replied:
done.

bzbarsky-apple replied:
What was done?  This is still not documenting why the removeDevice call is here.

nivi-apple replied:
i added the comment in both places. and looks like overwrote this. i will add the 2 lines back that's in https://github.com/project-chip/connectedhomeip/pull/33146/files#diff-ba829e4c29b4fae8a2eca2a13141a0a278b0d7db68b321e973cf7f26593638d5R3370 along with details about what causes the messed up state

nivi-apple replied:
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3020,6 +3021,120 @@ - (void)test032_MTRPathClassesEncoding
+        XCTAssert(attributeReportsReceived > 0);
+        for (NSDictionary<NSString *, id> * attributeDict in attributeReport) {
+            MTRAttributePath * attributePath = attributeDict[MTRAttributePathKey];
+            XCTAssert(attributePath != nil);
```
```suggestion
            XCTAssertNotNil(attributePath);
```
and similar things elsewhere: use `XCTAssertEqual()`, `XCTAssertEqualObjects`, etc.  Things like: `XCTAssertEqualObjects(attributePath.cluster, @(MTRClusterDescriptorID)`.

Except `MTRClusterDescriptorID` is deprecated, so don't use that.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3020,6 +3021,120 @@ - (void)test032_MTRPathClassesEncoding
+            XCTAssert(attributePath != nil);
+
+            XCTAssert(attributePath.cluster.unsignedLongValue == MTRClusterDescriptorID);
+            XCTAssert(attributePath.attribute.unsignedLongValue == MTRClusterDescriptorAttributePartsListID);
```
Again, no deprecated values, please.

nivi-apple replied:
I didn't realize this. Initially I used app::Clusters::Descriptor::Attributes::PartsList::Id but later saw usage of MTRClusterDescriptorAttributePartsListID and went with that. :(

nivi-apple replied:
fixed.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3020,6 +3021,120 @@ - (void)test032_MTRPathClassesEncoding
+            NSDictionary * dataValue = attributeDict[MTRDataKey];
+            XCTAssert(dataValue != nil);
+            NSArray<NSNumber *> * partsList = dataValue[MTRValueKey];
+            XCTAssert([partsList isEqual:(@[
```
Instead of two copies of this array, shouldn't we just have it in a variable somewhere?

nivi-apple replied:
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3020,6 +3021,120 @@ - (void)test032_MTRPathClassesEncoding
     XCTAssertEqualObjects(originalCommandPath, decodedCommandPath);
 }
 
+#ifdef DEBUG
```
```suggestion
```

There shouldn't be any DEBUG here.  The tests are always run in DEBUG.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -1090,10 +1108,42 @@ - (void)_reportAttributes:(NSArray<NSDictionary<NSString *, id> *> *)attributes
 {
     std::lock_guard lock(_lock);
 
+    [self _noteDeviceConfigurationChanged:attributeReport];
```
One more thing worth thinking about: in `_getAttributesToReportWithReportedValues` we are already doing the "did the value change from our cache" comparisons, right?   Perhaps in that code, if the value has changed, we should check whether the path is one of the ones we care about and set the flag?

That would incidentally also fix the fact that the "value has changed" check this PR implements is wrong, but the way the test is written, by injecting a value that does not match the values we would actually get passed to `_handleAttributeReport` does not catch that.

Which leads me to a request: can you please write a test that would actually test this properly?  Followup is OK for that, but the test here is clearly not it.

bzbarsky-apple replied:
And also: the test does not have tests that the "configuration changed" callback _is not called_ when the configuration has in fact not changed.  Those tests need to exist too, and would have caught the "value has changed" check being wrong.

nivi-apple replied:
This makes sense. added a test to the test017 to check if the device configuration changed callback is not called

nivi-apple committed:
Update src/darwin/Framework/CHIP/MTRDevice.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

nivi-apple committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

nivi-apple committed:
Addressed review comments

restyled-commits committed:
Restyled by clang-format

nivi-apple committed:
Merge branch 'master' into add_support_for_device_changed_notification_rebased

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -2515,6 +2565,11 @@ - (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSSt
             if (readCacheValueChanged && !expectedValue) {
                 shouldReportAttribute = YES;
+
+                _deviceConfigurationChanged = [self _isAttributeAffectingDeviceConfigurationChanged:attributePath];
```
This line looks wrong to me, both in terms of what it does and in terms of where it is located.  What is the logic we are trying to implement here?

Please write tests that would have caught at least the "what it does" bug; the other might be harder to test.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -1564,6 +1574,12 @@ - (void)test017_TestMTRDeviceBasics
         [resubscriptionGotReportsExpectation fulfill];
     };
 
+    // Set the onDeviceConfigurationChanged callback but do nothing. We shouldn't be called.
```
```suggestion
    // Make sure there are no device configuration changed notifications during this test.
    // There is nothing changing about the server that would lead to those.
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -1564,6 +1574,12 @@ - (void)test017_TestMTRDeviceBasics
     };
 
+    // Set the onDeviceConfigurationChanged callback but do nothing. We shouldn't be called.
+    __block BOOL wasOnDeviceConfigurationChangedCallbackCalled = false;
+    delegate.onDeviceConfigurationChanged = ^() {
+        wasOnDeviceConfigurationChangedCallbackCalled = true;
```
```suggestion
    __block BOOL wasOnDeviceConfigurationChangedCallbackCalled = NO;
    delegate.onDeviceConfigurationChanged = ^() {
        wasOnDeviceConfigurationChangedCallbackCalled = YES;
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
     XCTAssertEqualObjects(originalCommandPath, decodedCommandPath);
 }
 
+// Helper API to test if changes in an attribute with a path specified by endpointId, clusterId and attributeId affect
+// device configuration changed for a given MTRDevice. This API creates a fake attribute report for the given attribute path
```
```suggestion
// Helper API to test if changes in an attribute with a path specified by endpointId, clusterId and attributeId trigger
// device configuration changed callbacks for a given MTRDevice. This API creates a fake attribute report for the given attribute path
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+    for (NSDictionary<NSString *, id> * attributeDict in attributeReport) {
+        MTRAttributePath * attributePath = attributeDict[MTRAttributePathKey];
+        if (attributePath.attribute.unsignedLongLongValue == attributeId) {
+            testDataValue = attributeDict[MTRDataKey][MTRValueKey];
```
This doesn't make sense to me.  How do we know `attributeDict[MTRDataKey][MTRValueKey]` is an array?  In general I would expect it to _not_ be an array...

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+            MTRAttributePath * attributePath = attributeDict[MTRAttributePathKey];
+            XCTAssertNotNil(attributePath);
+
+            XCTAssert(attributePath.cluster.unsignedLongValue == clusterId);
+            XCTAssert(attributePath.attribute.unsignedLongValue == attributeId);
```
```suggestion
            XCTAssertEqualObjects(attributePath.cluster, @(clusterId));
            XCTAssertEqualObjects(attributePath.attribute, @(attributeId));
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+
+            NSDictionary * data = attributeDict[MTRDataKey];
+            XCTAssertNotNil(data);
+            XCTAssert([data[MTRDataVersionKey] isEqualToNumber:dataVersion]);
```
```suggestion
            XCTAssertEqualObjects(data[MTRDataVersionKey], dataVersion);
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+            NSArray<NSNumber *> * dataValue = data[MTRValueKey];
+            XCTAssertNotNil(dataValue);
+            XCTAssertNotNil(testDataValue);
+            XCTAssert([dataValue isEqual:testDataValue]);
```
```suggestion
            XCTAssertEqualObjects(dataValue, testDataValue);
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+                switch (attributePath.attribute.unsignedLongValue) {
+                case MTRAttributeIDTypeGlobalAttributeAttributeListID: {
+                    dataVersionForAttributeList = data[MTRDataVersionKey];
+                    ;
```
```suggestion
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+                }
+                case MTRAttributeIDTypeGlobalAttributeClusterRevisionID: {
+                    dataVersionForClusterRevision = data[MTRDataVersionKey];
+                    ;
```
```suggestion
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+                NSDictionary * data = attributeDict[MTRDataKey];
+                XCTAssertNotNil(data);
+                dataVersionForAcceptedCommandList = data[MTRDataVersionKey];
+                ;
```
```suggestion
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+                NSDictionary * data = attributeDict[MTRDataKey];
+                XCTAssertNotNil(data);
+                dataVersionForFeatureMap = data[MTRDataVersionKey];
+                XCTAssertNotNil(dataVersionForFeatureMap);
```
All these `XCTAssertNotNil(dataVersionForX)` should be happening in `onReportEnd`, no?  To make sure we actually got all the attributes we expected.

nivi-apple replied:
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+        }
+    } ];
+
+    [MTRDeviceTests runTestDeviceConfigurationChangedForAttribute:MTRAttributeIDTypeClusterDescriptorAttributePartsListID clusterId:MTRClusterIDTypeDescriptorID endpointId:@(0) device:device delegate:delegate dataVersion:testDataVersion attributeReport:attributeReport testcase:self];
```
Isn't that helper more like `checkAttributeReportTriggersConfigurationChanged` or so?

nivi-apple replied:
changed to checkAttributeReportTriggersConfigurationChanged

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+
+    [MTRDeviceTests runTestDeviceConfigurationChangedForAttribute:MTRAttributeIDTypeGlobalAttributeFeatureMapID clusterId:MTRClusterIDTypeGroupsID endpointId:@(0) device:device delegate:delegate dataVersion:testDataVersion attributeReport:attributeReport testcase:self];
+
+    // We need to remove the device here since the MTRDevice maintains its state and would interfere with the next test using MTRDevice.
```
Why would it interfere?  This should be clearer about why the state of this device is not ok.

nivi-apple replied:
sounds good. I will add more here its the state of the accessory reachable or not that causes the issues.

nivi-apple replied:
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3052,326 @@ - (void)test032_MTRPathClassesEncoding
+    // TODO: #33205 Ensure we have a clean slate w.r.t MTRDevice bfeore running each test.
+    [sController removeDevice:device];
+}
+
```
Missing tests:

1. Reports for some attribute that is not on our list do not trigger the callback.
2. Tests with reports for multiple attributes.  Might have caught the bug in the code, if done well enough.

nivi-apple replied:
sounds good. let me add those tests so i can figure out the bug

nivi-apple replied:
added the 2 tests - one for an attribute that doesn't trigger the callback and one with multiple reports.

nivi-apple committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

nivi-apple committed:
Update code to set _deviceConfigurationChanged correctly in _getAttributesToReportWithReportedValues

- Add more tests to verify that for attributes that don't affect device configuration changed, the
  delegate callback is not called. Also add tests for testing an attribute report having multiple
  attributes one of which causes device configuration changed.

restyled-commits committed:
Restyled by clang-format

nivi-apple committed:
Add logs to test the multiple attribute report

nivi-apple committed:
Fix the checkAttributeReportTriggersConfigurationChanged API and invert the expectation for device configuration changed when expectConfigurationChanged is set to NO

restyled-commits committed:
Restyled by clang-format

nivi-apple commented on src/darwin/Framework/CHIP/MTRDevice.h:
```diff
@@ -423,6 +423,15 @@ MTR_EXTERN NSString * const MTRDataVersionKey MTR_NEWLY_AVAILABLE;
+ *
+ * Device configuration is a set of functionality implemented by the device.
+ *
+ * The intention is that this is called whenever any of the things that usually don't change about a device might have changed.
```
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -2987,6 +3007,12 @@ - (void)test031_MTRDeviceAttributeCacheLocalTestStorage
+    // We need to remove the device here since the MTRDevice retains its reachable state. So if the next test needs to start with a clean state,
+    // it can't do that since the MTRDevice becomes reachable in the previous test. Since there are no changes detected in reachability,
+    // the onReachable callback to the delegate is not called.
+    // TODO: #33205 Ensure we have a clean slate w.r.t MTRDevice bfeore running each test.
```
```suggestion
    // TODO: #33205 Ensure we have a clean slate w.r.t MTRDevice before running each test.
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+        if (attributePath.attribute.unsignedLongLongValue == attributeId) {
+            NSDictionary * data = attributeDict[MTRDataKey];
+            XCTAssertNotNil(data);
+            testDataValue = data[MTRValueKey];
```
This does not make sense.  `testDataValue` is declared as `NSArray<NSDictionary<NSString *, id> *> *` but the `MTRValueKey` could be any type of value, no?

`testDataValue` should be declared as `id`.

nivi-apple replied:
will change everywhere I use MTRValueKey

nivi-apple replied:
This is my bad. yes it could be an unsigned int or an array, etc

nivi-apple replied:
done.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+            XCTAssertNotNil(data);
+            XCTAssertEqualObjects(data[MTRDataVersionKey], dataVersion);
+
+            NSArray<NSNumber *> * dataValue = data[MTRValueKey];
```
```suggestion
            id dataValue = data[MTRValueKey];
```
You have no idea what type it is unless you examine MTRTypeKey.  And you are not relying on the type anyway.

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+            XCTAssertEqualObjects(data[MTRDataVersionKey], dataVersion);
+
+            NSArray<NSNumber *> * dataValue = data[MTRValueKey];
+            XCTAssertNotNil(dataValue);
```
This does assume that none of the relevant attibutes have null as a value, yes?  OK for now, but might be worth a comment, because this could blow up depending on what attributes we test here.

nivi-apple replied:
added a comment

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+
+    [device unitTestInjectAttributeReport:attributeReport];
+
+    [testcase waitForExpectations:@[ gotAttributeReportExpectation, gotAttributeReportEndExpectation, deviceConfigurationChangedExpectation ] timeout:kTimeoutInSeconds];
```
Huh, so waiting on an inverted expectation works fine?  Good to know!

nivi-apple replied:
yup I was going to ask you yesterday if thats ok and found that it is during testing. good to know

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+        }
+    } ];
+
+    // Test attribute path - endpointId = 0, clusterId = descriptor, attributeId = server list.
+    [MTRDeviceTests checkAttributeReportTriggersConfigurationChanged:MTRAttributeIDTypeClusterDescriptorAttributeDeviceTypeListID clusterId:MTRClusterIDTypeDescriptorID endpointId:@(0) device:device delegate:delegate dataVersion:dataVersionForDeviceTypesList attributeReport:attributeReport testcase:self expectConfigurationChanged:YES];
+
+    dataVersionForServerList = [NSNumber numberWithUnsignedLongLong:(dataVersionForServerList.unsignedLongLongValue + 1)];
```
```suggestion
    [MTRDeviceTests checkAttributeReportTriggersConfigurationChanged:MTRAttributeIDTypeClusterDescriptorAttributeDeviceTypeListID clusterId:MTRClusterIDTypeDescriptorID endpointId:@(0) device:device delegate:delegate dataVersion:dataVersionForDeviceTypesList attributeReport:attributeReport testcase:self expectConfigurationChanged:YES];

    // Test attribute path - endpointId = 0, clusterId = descriptor, attributeId = server list.
    dataVersionForServerList = [NSNumber numberWithUnsignedLongLong:(dataVersionForServerList.unsignedLongLongValue + 1)];
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+    // We need to remove the device here since the MTRDevice retains its reachable state. So if the next test needs to start with a clean state,
+    // it can't do that since the MTRDevice becomes reachable in the previous test. Since there are no changes detected in reachability,
+    // the onReachable callback to the delegate is not called.
+    // TODO: #33205 Ensure we have a clean slate w.r.t MTRDevice bfeore running each test.
```
```suggestion
    // TODO: #33205 Ensure we have a clean slate w.r.t MTRDevice before running each test.
```

bzbarsky-apple commented on src/darwin/Framework/CHIPTests/MTRDeviceTests.m:
```diff
@@ -3033,6 +3059,420 @@ - (void)test032_MTRPathClassesEncoding
+    [device unitTestInjectAttributeReport:attributeReport];
+    [self waitForExpectations:@[ gotAttributeReportWithMultipleAttributesExpectation, gotAttributeReportWithMultipleAttributesEndExpectation, deviceConfigurationChangedExpectationForAttributeReportWithMultipleAttributes ] timeout:kTimeoutInSeconds];
+
+    // We need to remove the device here since the MTRDevice retains its reachable state. So if the next test needs to start with a clean state,
+    // it can't do that since the MTRDevice becomes reachable in the previous test. Since there are no changes detected in reachability,
+    // the onReachable callback to the delegate is not called.
```
```suggestion
    // We need to remove the device here, because we injected data into its attribute cache
    // that does not match the actual server.
```

nivi-apple committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

nivi-apple committed:
Addressed comments
